Google.Api.Generator.Utils/SourceFileContext.cs

```diff
@@ -270,58 +294,107 @@ string ImportAliased(string ns)
+                    void Add(IDictionary<string, HashSet<string>> tNamespaces, string t, string ns)
+                    {
+                        if (!tNamespaces.TryGetValue(t, out HashSet<string> namespaces))
                         {
-                            // We are just going to remove one collision out of two, so
-                            // this type won't be colliding at all anymore.
-                            collidingTypes.Remove(t);
+                            namespaces = tNamespaces[t] = new HashSet<string>();
                         }
-                        else
+                        namespaces.Add(ns);
+                    }
+
+                    void RemoveNamespaceFromColliding(string ns)
+                    {
+                        foreach(var tNs in collidingTypes.Where(ctNs => ctNs.Value.Contains(ns)).ToList())
+                        {
+                            RemoveTypeFromColliding(tNs.Key, ns);
+                        }
+                    }
+
+                    bool RemoveTypeFromColliding(string t, string ns)
+                    {
+                        if (collidingTypes.TryGetValue(t, out var collidingTypeNamespaces) && collidingTypeNamespaces.Contains(ns))
                         {
-                            // There are more than two collisions so we just remove
-                            // the collision introduced by this namespace.
-                            collidingTypeNamespaces.Remove(ns);
+                            if (collidingTypeNamespaces.Count == 2 || // If we remove one of the remaining two collisions...
+                                // or if after removal the remaining collisions are between unseen imported types...
+                                !collidingTypeNamespaces.Any(ctNs => ctNs != ns && seenTypes.Any(typ => typ.Name == t && typ.Namespace == ctNs)))
+                            {
+                                // ... then this type is not colliding anymore
+                                collidingTypes.Remove(t);
+                            }
+                            else
+                            {
+                                // Even after removal of this collision, there are remaining collisions from seen types
+                                // with this name, so we just remove this particular collision.
+                                collidingTypeNamespaces.Remove(ns);
+                            }
+                            return true;
                         }
+                        return false;
                     }
                 }
 
-                private TypeAliaser(Dictionary<string, string> aliasedNamespaces, HashSet<string> unaliasedNamespaces, Dictionary<string, HashSet<string>> aliasedTypes) =>
-                    (_aliasedNamespaces, _unaliasedNamespaces, _aliasedTypes) = (aliasedNamespaces, unaliasedNamespaces, aliasedTypes);
+                private TypeAliaser(
```

Comment:
Can we pull each parameter onto a new line here? It's getting rather hard to read.
For the assignment, I'd just use two lines (one for targets, one for sources)

Reply:
Done